art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -0,0 +1,389 @@
+        # Choose samples to poison.
+        x_train = np.copy(x_train)
+        y_train = np.copy(y_train)
+        x_trigger = self.apply_trigger_patch(x_trigger)
+        if len(np.shape(y_trigger)) == 2:  # dense labels
+            classes_target = set(np.argmax(y_trigger, axis=-1))
+        else:  # sparse labels
+            classes_target = set(y_trigger)
+        num_poison_samples = int(self.percent_poison * len(x_train))
+
+        # Try poisoning num_trials times and choose the best one.
+        best_B = np.finfo(np.float32).max  # pylint: disable=C0103
+        best_x_poisoned = None
+        best_indices_poison = None
+
+        if len(np.shape(y_train)) == 2:
+            y_train_classes = np.argmax(y_train, axis=-1)
+        else:
+            y_train_classes = y_train
+        for _ in trange(self.max_trials):
+            if self.selection_strategy == "random":
+                self.indices_poison = np.random.permutation(
+                    np.where([y in classes_target for y in y_train_classes])[0]
+                )[:num_poison_samples]
+            else:
+                self.indices_poison = self.select_poison_indices(
+                    self.substitute_classifier, x_train, y_train, num_poison_samples
+                )
+            x_poison = x_train[self.indices_poison]
+            y_poison = y_train[self.indices_poison]
+            initializer(x_trigger, y_trigger, x_poison, y_poison)
+            original_epochs = self.max_epochs
+            if self.model_retrain:
+                retrain_epochs = self.max_epochs // self.retraining_factor
+                for i in range(self.retraining_factor):
+                    if i == self.retraining_factor - 1:
+                        self.max_epochs = original_epochs - retrain_epochs * i
+                        x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+                    else:
+                        self.max_epochs = retrain_epochs
+                        x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+                        self.model_retraining(x_poisoned)
+            else:
+                x_poisoned, B_ = poisoner(x_poison, y_poison)  # pylint: disable=C0103
+            finish_poisoning()
+            B_ = np.mean(B_)  # Averaging B losses from multiple batches.  # pylint: disable=C0103
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = self.indices_poison
+
+        if self.verbose > 0:
+            print("Best B-score:", best_B)
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train
+
+    def get_poison_indices(self):
+        """
+        :return: indices of best poison index
+        """
+        return self.indices_poison
+
+    def model_retraining(self, poisoned_samples):
```

Comment:
Is this method also used for TensorFlow in a future release? If yes, should we for now raise a warning of the model is in TensorFlow?

Reply:
Sure

Reply:
Done